dependency/vault_pki.go

```diff
@@ -66,40 +64,68 @@ func (d *VaultPKIQuery) Fetch(clients *ClientSet, opts *QueryOptions,
 	default:
 	}
 
-	var encPEM []byte
-	var err error
-	if encPEM, err = os.ReadFile(d.filePath); err != nil || len(encPEM) == 0 {
-		// no need to write to file as it is the template dest
-		encPEM, err = d.fetchPEM(clients)
-	}
-	if err != nil {
-		return nil, nil, errors.Wrap(err, d.String())
-	}
+	needsRenewal := fmt.Errorf("needs renewal")
+	getPEM := func(renew bool) ([]byte, error) {
+		var encPEM []byte
+		var err error
+		encPEM, err = os.ReadFile(d.filePath)
+		if renew || err != nil || len(encPEM) == 0 {
+			encPEM, err = d.fetchPEM(clients)
+			// no need to write cert to file as it is the template dest
+		}
+		if err != nil {
+			return nil, err
+		}
 
-	block, cert, err := getCert(encPEM)
-	if err != nil {
-		return nil, nil, err
+		block, cert, err := getCert(encPEM)
+		if err != nil {
+			return nil, err
+		}
+
+		if sleepFor, ok := goodFor(cert); ok {
+			d.sleepCh <- sleepFor
+			return pem.EncodeToMemory(block), nil
+		}
+		return []byte{}, needsRenewal
 	}
 
-	if sleepFor, ok := goodFor(cert); ok {
-		d.sleepCh <- sleepFor
-		return respWithMetadata(string(pem.EncodeToMemory(block)))
+	pemBytes, err := getPEM(false)
+	switch err {
+	case nil:
+	case needsRenewal:
+		pemBytes, err = getPEM(true)
+		if err != nil {
+			return nil, nil, err
+		}
+	default:
+		return nil, nil, err
 	}
-	d.pemBlock = nil
-	return "", &ResponseMetadata{
-		LastContact: 0,
-		LastIndex:   0,
-	}, nil
+	return respWithMetadata(string(pemBytes))
 }
 
-// returns time left in 90% of the original lease
-// boolean returns false if time has already past
+// returns time left in ~90% of the original lease and a boolean
+// that returns false if cert needs renewing, true otherwise
 func goodFor(cert *x509.Certificate) (time.Duration, bool) {
 	start, end := cert.NotBefore.Unix(), cert.NotAfter.Unix()
-	// use 90% of the cert duration
-	goodfor := ((end - start) * 9) / 10
-	sleepFor := time.Until(time.Unix(start+goodfor, 0))
-	return sleepFor, sleepFor > 0
+	now := time.Now().UTC().Unix()
+	if end <= now || start >= now {
```

Comment:
The `start >= now` case can fail mysteriously sometimes. Better to add some fuzziness to it, I think.

Reply:
I thought this was supposed to be dealt with on the Vault side with the not_before_duration time padding. It defaults to starting the NotBefore 30s in the past and figured this was where they were supposed to deal with these issues. Seems like making if fuzzy would just confuse things as it could fail sometimes and not others.

Reply:
I was thinking of the case where someone configures NotBefore to be 1s (or 0s), and then something weird happens with the clock (NTP sync runs, weird daylight savings time problem, leap second, etc.) and `time.Now()` goes backwards.

Reply:
If they use a `1s` NotBefore then it seems like they would want to know about skew issues as they are setting the value used for skew management to it's lowest setting. (BTW you can't set it to 0 as that is the zero value and triggers the default 30s).

Just seems to me that we are adding some in-determinism to the code (bad in my book) to handle something that is already handled deterministically (ie. better) someplace else.

Reply:
I'd say comparing timestamps is never as deterministic as we want it to be. :) I'd almost be in favor of just removing the `start >= now` case?

Reply:
I'm OK with just deleting it. It was more a matter of completeness than anything. I didn't have it originally.

Reply:
Deleted it.